feat: migrate editor to prosekit

ocavue opened the pull request:
## What does this PR do?

Migrate the text editor from Lexical to ProseKit. 

This is a draft PR and not all features are finished. See the list below for details. 

## Feature list

In this section I listed all features that I implemented in the PR. Each feature has a screen recording to show its behavior. Currently I show both the new ProseKit editor and the old Lexical editor for debugging and testing purpose. I will remove the Lexical editor once this PR is good enough. 

### Style text by keyboard shortcut (e.g. `Command B`)

<details><summary>Video</summary>

https://gist.github.com/assets/24715727/6f062e19-d510-4a42-a07e-a08e4a2383da

</details>

### Style text by input rule (e.g. `**bold**`)

You can see a bug in the Lexical editor when typing `~~delete~~`. This migration fixes it. 
 
<details><summary>Video</summary>

https://gist.github.com/assets/24715727/5037016d-64f5-4d2b-ba06-24f4959674ca

</details>


### Show a floating inline menu when text is selected 

This is a new feature. 

<details><summary>Video</summary>
 
https://gist.github.com/assets/24715727/eba1ae74-05ed-45fd-8d14-d48aea116c06

</details>

### Mention picker 


<details><summary>Video</summary>
 
https://gist.github.com/assets/24715727/3f433ad1-2e7a-4395-ac3a-35bfc30eed7f

</details>


### Emoji picker 


<details><summary>Video</summary>

https://gist.github.com/assets/24715727/a1b4ee9c-4a2d-4daf-99da-ed86f2b7db3a

</details>

### Link matching


<details><summary>Video</summary>

(HTTP links)
 
https://gist.github.com/assets/24715727/18dade66-b64f-42f5-ac30-35e4977ca4cf
 
(Email links)
 
https://gist.github.com/assets/24715727/50647336-5e88-4e2e-a263-158303602ae6

</details>

There are various improvements for link matching compared to the previous Lexical editor. Now the link matching is more robust and it has similar behavior as Twitter's link matching. 


### Hashtag and Cashtag 


<details><summary>Video</summary>

 https://gist.github.com/assets/24715727/fb043312-f480-40f5-a8e0-a55827449797

</details>

Cashtag was never implemented in the Lexical editor, so this would be a new feature.

### Drag and drop to insert images 

<details><summary>Video</summary>

https://gist.github.com/assets/24715727/fe4e35bb-c60f-458f-bc70-b69c80511c10

</details>


## Related issues

Fixes # (issue)

HEY-106

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Enhancement (non-breaking small changes to existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Explanation of the changes

bigint commented:
https://github.com/heyxyz/hey/assets/69431456/0df57f5a-1885-4374-8def-0d51f4bc9e58

1. We can reset the mentions list when entered or backspaced till `@`, I see stale list coming even after selecting it
2. We don't want to show an empty state if the mention list is empty, can just hide it

bigint commented:
https://github.com/heyxyz/hey/assets/69431456/363c97c5-2ad7-4d49-bc7d-2cc62b75f277

I can't type hashtags after mentioning the user

bigint commented:
https://github.com/heyxyz/hey/assets/69431456/9b424b1e-2451-4f84-b56a-489ebf67aad4

The emoji picker is not appending/inserting emojis

bigint commented:
https://github.com/heyxyz/hey/assets/69431456/57eae837-fe29-450f-b9e9-af5bc75814e5

Emoji picker popping up when manually typing the links

bigint commented:
Uploading Screen Recording 2024-04-29 at 4.36.53 PM.mov…

Text autofill via url not working

bigint commented on apps/web/package.json:
```diff
@@ -55,11 +55,13 @@
     "graphql": "^16.8.1",
     "idb-keyval": "^6.2.1",
     "lexical": "^0.14.5",
+    "lucide-react": "^0.376.0",
```
We can use heroicons which we already have instead of lucide

bigint commented on apps/web/src/components/Composer/TextEditor/EmojiPicker.tsx:
```diff
@@ -0,0 +1,80 @@
+
+import { useEmojiQuery } from './useEmojiQuery';
+
+function EmojiEmpty() {
```
We always prefer using arrow function/comps everywhere 🙇🏼

bigint commented on apps/web/package.json:
You can remove lexical completely 🙇🏼

ocavue replied:
I just found the draft feature, which is hidden under the `isPro` flag, when I tried to remove lexical. I will first implement the draft (i.e. load markdown string as editable content) then I will remove lexical.

bigint commented on apps/web/src/components/Composer/TextEditor/extension.ts:
```diff
@@ -0,0 +1,88 @@
+import { defineUnderline } from 'prosekit/extensions/underline';
+import { defineVirtualSelection } from 'prosekit/extensions/virtual-selection';
+
+const EMAIL_MATCHER =
```
We can remove email support 🙇🏼, it's not required IMO

bigint commented on apps/web/src/components/Composer/TextEditor/index.tsx:
```diff
@@ -0,0 +1,6 @@
+import dynamic from 'next/dynamic';
+
+// Don't render the editor on the server because some parts of it depend on the browser API.
+const DynamicTextEditor = dynamic(() => import('./TextEditor'), { ssr: false });
```
We can remove dynamic loading, cuz it causes some CLS! We don't have any server components, we added is server check globally in the app

ocavue replied:
I only dynamic load floating menus now so it won't cause CLS anymore.

bigint commented on apps/web/src/components/Composer/TextEditor/useContentChange.ts:
```diff
@@ -0,0 +1,27 @@
+
+import { markdownFromHTML } from './markdown';
+
+export function useContentChange(editor: Editor<TextEditorExtension>) {
```
Arrow function

bigint commented on apps/web/src/components/Composer/TextEditor/useEmojiQuery.ts:
```diff
@@ -0,0 +1,35 @@
+  const [loaded, setLoaded] = useState(false);
+
+  const fetchEmojis = async () => {
+    const res = await fetch(`${STATIC_ASSETS_URL}/emoji.json`);
```
You can use axios, we avoid using fetch 🙇🏼

bigint commented on apps/web/src/components/Composer/TextEditor/useEmojiQuery.ts:
```diff
@@ -0,0 +1,35 @@
+
+  useEffect(() => {
+    if (load && !loaded) {
+      fetchEmojis();
```
You can wrap this to use tanstack query `useQuery`

ocavue commented on apps/web/package.json:
```diff
@@ -81,8 +86,10 @@
+    "@egoist/tailwindcss-icons": "^1.7.4",
     "@hey/config": "workspace:*",
     "@hey/types": "workspace:*",
+    "@iconify-json/lucide": "^1.1.183",
```
Heroicons doesn't have text editing icons like bold, italic, underline.

ocavue commented:
![Uploading image.png…]()

> Text autofill via url not working

The video is not been uploaded.

ocavue commented:
> Emoji picker popping up when manually typing the links

Fixed 

> The emoji picker is not appending/inserting emojis

Fixed 

> I can't type hashtags after mentioning the user

Fixed 

> We can reset the mentions list when entered or backspaced till @, I see stale list coming even after selecting it

Fixed 

> We don't want to show an empty state if the mention list is empty, can just hide it

Fixed

bigint commented:
@mention should be @lens/mention when posted else it will not be rendered, but it should be shown as @mention on the composer

https://web-git-fork-ocavue-ocavue-prosekit-heyxyz.vercel.app/posts/0x05-0x0a-DA-994917ed

bigint commented:
> The video was not been uploaded.

https://github.com/heyxyz/hey/assets/69431456/bcefb40a-ff78-485f-a195-c0d3e313a312

ocavue commented:
> @mention should be @lens/mention when posted else it will not be rendered, but it should be shown as @mention on the composer

Fixed. 

<img width="839" alt="image" src="https://github.com/heyxyz/hey/assets/24715727/4ce2ffe5-9dcd-4870-9a58-d8fa208a8427">



---


> Text autofill via url not working

Fixed. 

![CleanShot 2024-05-03 at 01 30 18@2x](https://github.com/heyxyz/hey/assets/24715727/8cebad6e-f6ad-4ace-b785-2a8e75dceb67)

ocavue commented:
@bigint I addressed all your previous review comments. Let me know if there's anything else I can do.

bigint commented:
Thanks @ocavue lemme to the review ASAP 🙇🏼

bigint commented:
https://web-git-fork-ocavue-ocavue-prosekit-heyxyz.vercel.app/posts/0x05-0x0a-DA-b9df8832

I think we should pefix with https:// if user gave an empty link

bigint commented on apps/web/src/components/Composer/TextEditor/EmojiPicker.tsx:
```diff
@@ -0,0 +1,73 @@
+  );
+};
+
+const EmojiPicker = () => {
```
It's not working for me until i type the third character

bigint replied:
We wan't something similar like github editor 🙇🏼 has with emoji search

ocavue replied:
Done.

<img width="383" alt="image" src="https://github.com/heyxyz/hey/assets/24715727/7a1b364b-ade3-4083-a3a3-748025ddf31b">

I thought you want to start showing emoji when you have some characters typed based on the existing behavior in another emoji picker ([code link](https://github.com/ocavue/hey/blob/680f93876e2fd56b6e4c8fe1ae89dfd2fa77703e/apps/web/src/components/Shared/EmojiPicker/List.tsx#L36)). I've changed it and now it works exactly like Github.

bigint commented on apps/web/src/components/Composer/TextEditor/MentionPicker.tsx:
```diff
@@ -0,0 +1,96 @@
+      )}
+      offset={10}
+      onQueryChange={setQueryString}
+      regex={/@\w*$/}
```
can we reuse or move to `regex.ts` file?

ocavue replied:
Done

bigint commented on apps/web/src/components/Composer/TextEditor/Toggle.tsx:
```diff
@@ -0,0 +1,33 @@
+  return (
+    <Tooltip content={tooltip} placement="top">
+      <button
+        className="outline-unset focus-visible:outline-unset flex items-center justify-center rounded-md bg-transparent p-2 font-medium transition hover:bg-zinc-100 focus-visible:ring-2 focus-visible:ring-zinc-900 disabled:pointer-events-none disabled:opacity-50 hover:disabled:opacity-50 data-[state=on]:bg-gray-200/60 dark:hover:bg-zinc-800 dark:focus-visible:ring-zinc-300 dark:data-[state=on]:bg ...
```
we can use gray instead of zinc 🙇🏼  everywhere

ocavue replied:
done

bigint commented:
Rest looks good!

small changes @ocavue

ocavue commented on apps/web/src/components/Composer/NewPublication.tsx:
```diff
@@ -241,7 +236,7 @@ const NewPublication: FC<NewPublicationProps> = ({ publication }) => {
     // Track in leafwatch
     const eventProperties = {
-      comment_on: isComment ? publication.id : null,
+      comment_on: isComment ? publication?.id : null,
```
`<NewPublication>` accepts an optional `publication` prop. See the following two usages:

- https://github.com/ocavue/hey/blob/99f9c521dc993b216b443605ffdf02a1e75094b1/apps/web/src/components/Publication/index.tsx#L158

- https://github.com/ocavue/hey/blob/99f9c521dc993b216b443605ffdf02a1e75094b1/apps/web/src/components/Shared/GlobalModals.tsx#L133


Thus, the type of `publication` should be `publication?: MirrorablePublication`. TypeScript didn't complain about this issue before my PR because the previous `withLexicalContext` HOC just marks the component props [as `any`](https://github.com/ocavue/hey/blob/99f9c521dc993b216b443605ffdf02a1e75094b1/apps/web/src/components/Shared/Lexical/withLexicalContext.tsx#L27). I removed `withLexicalContext` and also correct this type error.

ocavue commented on apps/web/src/components/Composer/NewPublication.tsx:
```diff
@@ -551,7 +539,7 @@ const NewPublication: FC<NewPublicationProps> = ({ publication }) => {
         />
       ) : null}
-      <Editor />
+      <TextEditor defaultMarkdown={publicationContent} editorRef={editorRef} />
```
> Let's name it itself

Sorry I don't understand this comment

bigint replied:
Oh, my wrong!

Lets name it just `<Editor />` instead of `<TextEditor />`

ocavue replied:
Done

ocavue commented on apps/web/src/components/Composer/Actions/DraftSettings/index.tsx:
```diff
@@ -28,7 +30,7 @@ const DraftSettings: FC = () => {
         title="Drafts"
       >
-        <List setShowModal={setShowModal} />
+        <List editorRef={editorRef} setShowModal={setShowModal} />
```
I removed `React.useImperativeHandle` and `editorRef`. I'm using a context to pass editor handle now, which is similar to what we used to have when using Lexical. 

See `TextEditor/EditorHandle.tsx` for more details.

ocavue commented on apps/web/src/components/Composer/Actions/DraftSettings/index.tsx:
```diff
@@ -1,11 +1,13 @@
 import List from './List';
 
-const DraftSettings: FC = () => {
+const DraftSettings: FC<{ editorRef: EditorRef }> = ({ editorRef }) => {
```
I've removed `editorRef` so this file is not touched anymore.

ocavue commented on apps/web/src/components/Composer/Actions/DraftSettings/List.tsx:
```diff
@@ -17,18 +16,17 @@ import { useCollectModuleStore } from 'src/store/non-persisted/publication/useCo
 import { usePublicationStore } from 'src/store/non-persisted/publication/usePublicationStore';
 
 interface ListProps {
+  editorRef: EditorRef;
```
I've removed `editorRef`.

ocavue commented:
> I think we should pefix with https:// if user gave an empty link

Fixed

ocavue commented:
@bigint I've addressed all your comments above.

bigint commented:
https://github.com/heyxyz/hey/assets/69431456/f8754c7b-8159-4aa5-9669-c27aa7f360e8

@ocavue one thing i found while testing is the content is not getting cleared after posting

ocavue commented:
> one thing i found while testing is the content is not getting cleared after posting

I cannot reproduce this issue. The web page will reload after I approved the signature request in my MetaMask Chrome extension. 

https://github.com/heyxyz/hey/assets/24715727/53d71a8f-3d5f-42f4-9044-2f1aef956a5b

Although I cannot reproduce it, I'll check the code and see if there are potential bugs that cause it.

ocavue commented:
> one thing i found while testing is the content is not getting cleared after posting

I cannot reproduce this issue. The web page will reload after I approved the signature request in my MetaMask Chrome extension. 

https://github.com/heyxyz/hey/assets/24715727/53d71a8f-3d5f-42f4-9044-2f1aef956a5b

Although I cannot reproduce it, I'll check the code and see if there are potential bugs that causes it.

bigint commented:
https://github.com/heyxyz/hey/assets/69431456/3fe1d06e-785c-49cd-b593-7020dcdd0012

Getting this now 🙇🏼 

Try commenting on this post https://web-git-fork-ocavue-ocavue-prosekit-heyxyz.vercel.app/posts/0x05-0x0c-DA-c5acc0cc

ocavue commented:
> Getting this now 🙇🏼

If the document is large, the serialization (editor data structure => HTML string => Markdown string) could be slow. That's way I add a 500ms debounce for serialization ([code link](https://github.com/ocavue/hey/blob/70dd889d53aa3379f60e1adb7f032203132abdf8/apps/web/src/components/Composer/Editor/useContentChange.ts#L24)). 

The downside of this debounce is that, if you type something very fast and press the submit button immediately, the Markdown string is not ready yet. That's why you saw this error.

Let me think what's the best way to resolve it (and fit existing coding pattern).

bigint commented:
@ocavue One last thing i found is the mention dropdown not autoselecting the 0th index

https://github.com/heyxyz/hey/assets/69431456/4129cd46-924b-4cc0-b362-9f949d26eb36

ocavue commented:
> The avatar in the composer causing some CLS

Fixed. This issue is caused by the overflowing when there is a long word (like a long link) in the editor. 

## Before 

<img width="622" alt="Screenshot 2024-05-05 at 17 39 38" src="https://github.com/heyxyz/hey/assets/24715727/b342daaa-ca69-49ad-8e06-c4f8c01ba521">

## After 

<img width="619" alt="Screenshot 2024-05-05 at 17 43 43" src="https://github.com/heyxyz/hey/assets/24715727/26bcb446-b647-41c3-9cbd-124eadc8c1e1">

bigint commented:
Also for mention, lets not limit to first 3 chars, we can start searching from the first character similar to twitter or github

bigint commented:
Hovering profile on mention dropdown should show cursor-pointer, RN it shows text selection cursor

bigint commented:
https://github.com/heyxyz/hey/assets/69431456/dee8aea8-2d71-4300-bf84-e914a06dffc9

it should show the username and not the profile name 🙇🏼

bigint commented:
**LGTM so far** :shipit: 

![LGTM](https://media3.giphy.com/media/dASc6rD8EOXEQ/giphy.gif)

ocavue commented:
> One last thing i found is the mention dropdown not autoselecting the 0th index

Fixed. 

<img width="392" alt="image" src="https://github.com/heyxyz/hey/assets/24715727/0d0a7cf1-6ebe-4528-b134-75bd495344de">
